rust/perspective-python/src/client/client_async.rs

```diff
@@ -33,60 +33,75 @@ use super::update_data::UpdateDataExt;
 use super::{pandas, polars, pyarrow};
 use crate::py_err::{PyPerspectiveError, ResultTClientErrorExt};
 
+#[pyclass(module = "perspective")]
 #[derive(Clone)]
-pub struct PyClient {
+pub struct AsyncClient {
     pub(crate) client: Client,
     loop_cb: Arc<RwLock<Option<Py<PyAny>>>>,
     close_cb: Option<Py<PyAny>>,
 }
 
-impl PyClient {
+impl AsyncClient {
+    pub fn new_from_client(client: Client) -> Self {
+        AsyncClient {
+            client,
+            loop_cb: Arc::default(),
+            close_cb: None,
+        }
+    }
+}
+
+#[pymethods]
+impl AsyncClient {
+    #[new]
+    #[pyo3(signature=(handle_request, handle_close=None))]
     pub fn new(handle_request: Py<PyAny>, handle_close: Option<Py<PyAny>>) -> Self {
+        let handle_request = Arc::new(handle_request);
         let client = Client::new_with_callback({
             move |msg| {
-                clone!(handle_request);
+                let handle_request = handle_request.clone();
                 Box::pin(async move {
-                    Python::with_gil(move |py| {
-                        handle_request.call1(py, (PyBytes::new_bound(py, msg),))
-                    })?;
+                    if let Some(fut) = Python::with_gil(move |py| -> PyResult<_> {
+                        let ret = handle_request.call1(py, (PyBytes::new(py, msg),))?;
+                        if isawaitable(ret.bind(py)).unwrap_or(false) {
```

Comment:
Crucial change here: if calling the Python callback returns a coroutine (i.e. if the callback is a coroutine function), then convert the coro into a Rust future and then await it (line 72)